Add support for `ignoreOtherConfiguration` setting option

esanzgar opened the pull request:
As part of https://github.com/hypothesis/viahtml/pull/192 we decided to introduce a new `ignoreOtherConfiguration` option in the hypothesis settings.

On the first commit, I added support on non-JSON configuration (`window.hypothesisConfig()`)

On the second commit, I added a test case to  check the effect on this new option.

robertknight commented on src/annotator/config/settings.js:
```diff
@@ -20,9 +21,21 @@ import isBrowserExtension from './is-browser-extension';
-  const jsonConfigs = parseJsonConfig(window_.document);
+  // Prioritize the `window.hypothesisConfig` function over the JSON format
+  // Via uses `window.hypothesisConfig` and makes it non-configurable and non-writable.
+  // In addition, `via` sets the `ignoreOtherConfiguration` option to prevent configuration merging.
```
```suggestion
  // In addition, Via sets the `ignoreOtherConfiguration` option to prevent configuration merging.
```

robertknight reviewed (commented):
The changes related to `window.hypothesisConfig` look good. Regarding JSON script tags, I think we should leave that out for the moment. There are multiple ways that we might want the flag to work in that context (apply only the configuration from the JSON tag containing `ignoreOtherConfiguration`, apply the cumulative configuration up until the `ignoreOtherConfiguration` flag is encountered as you've done here) and I think it would be wise to wait for a concrete use case to inform which of them to implement.

esanzgar commented:
> Regarding JSON script tags, I think we should leave that out for the moment.

I agree with you. I dropped the second commit and instead I added a test scenario to check this new option.

robertknight commented on dev-server/documents/html/configurations.mustache:
```diff
@@ -0,0 +1,48 @@
+  <body>
+    <h1>Ignore configuration test</h1>
+    <p>
+      `viahtml` sets the `ignoreOtherConfiguration` option on
+      `window.hypotesysConfig`. This option disregards other JSON hypothesis
```
```suggestion
      `window.hypothesisConfig`. This option disregards other sources of Hypothesis
```

I changed "JSON" to "sources" here since the point is that it doesn't matter what kind of configuration is used elsewhere, it just ignores them all.

robertknight replied:
```suggestion
      `viahtml` sets the `ignoreOtherConfiguration` option in the result of
      `window.hypothesisConfig()`. This option disregards other JSON hypothesis
```

To be precise, `hypothesisConfig` is a function that returns an object, rather than an object itself as the previous wording might have suggested.

robertknight commented on dev-server/documents/html/configurations.mustache:
```diff
@@ -0,0 +1,48 @@
+    <p>
+      `viahtml` sets the `ignoreOtherConfiguration` option on
+      `window.hypotesysConfig`. This option disregards other JSON hypothesis
+      configurations that the host page could include.
```
```suggestion
      configuration that the host page could include.
```

robertknight commented on dev-server/documents/html/configurations.mustache:
```diff
@@ -0,0 +1,48 @@
+      configurations that the host page could include.
+    </p>
+    <p>
+      In addition, `viahtml` makes `window.hypotesysConfig` non-configurable and
```
```suggestion
      In addition, `viahtml` makes `window.hypothesisConfig` non-configurable and
```

robertknight commented on dev-server/documents/html/configurations.mustache:
```diff
@@ -0,0 +1,48 @@
+      behave on `via` (PDF or html).
+    </p>
+    <p>
+      If this test works, it should make the sidebar to remain closed and to
+      have the classic theme. The configurations `"openSidebar" : true` and
```
```suggestion
      If this test works, it should make the sidebar remain closed and
      use the classic theme. The options `"openSidebar" : true` and
```

robertknight commented on src/annotator/config/settings.js:
```diff
@@ -20,9 +21,21 @@ import isBrowserExtension from './is-browser-extension';
+  let jsonConfigs;
+  if (
+    configFuncSettings.hasOwnProperty('ignoreOtherConfiguration') &&
+    toBoolean(configFuncSettings.ignoreOtherConfiguration) === true
```
Assuming that `toBoolean` always returns a boolean as the name implies the `=== true` part is not strictly needed. Also it it casts undefined => boolean then the `hasOwnProperty` part is not strictly needed either.

esanzgar replied:
That's true. The only reason to use `hasOwnProperty` would be to avoid prototype pollution.

robertknight commented on dev-server/documents/html/configurations.mustache:
```diff
@@ -0,0 +1,48 @@
+  <body>
+    <h1>Ignore configuration test</h1>
+    <p>
+      `viahtml` sets the `ignoreOtherConfiguration` option on
```
The dev server doesn't process markdown in in templates so you'll need to use `<code>someSymbol</code` rather than backticks.

robertknight replied:
Including this context about the purpose and usage of the option is very helpful.

robertknight commented on dev-server/documents/html/configurations.mustache:
```diff
@@ -0,0 +1,48 @@
+<!DOCTYPE html>
```
This template has a rather generic name (`configurations.mustache`). I suggest something like `ignore-other-configuration.mustache` would leave room for other configuration-related tests in future.

esanzgar replied:
Sure.

robertknight commented on dev-server/templates/index.mustache:
```diff
@@ -24,6 +24,7 @@
     <li><a href="/document/parent-frame">Annotatable iframe test</a></li>
     <li><a href="/document/sidebar-external-container">Sidebar in external container test</a></li>
     <li><a href="/document/multi-frames">Multi-frame test</a></li>
+    <li><a href="/document/configurations">Ignore configuration test</a></li>
```
```suggestion
    <li><a href="/document/configurations"><code>ignoreOtherConfiguration</code> test</a></li>
```

I think it might be hard for a new developer to understand what "ignore configuration" means. Naming it the way I've suggested here would hopefully make it more obvious that it refers to a specific configuration option.

robertknight commented on dev-server/documents/html/configurations.mustache:
```diff
@@ -0,0 +1,48 @@
+    </p>
+    <p>
+      In addition, `viahtml` makes `window.hypotesysConfig` non-configurable and
+      non-writable. These two settings together makes the client consistenly
```
```suggestion
      non-writable. These two settings together makes the client consistently
```

robertknight reviewed (changes requested):
This looks good from a functional point of view. I requested some corrections on the wording and formatting of the test page.

esanzgar commented:
@robertknight Thanks for your suggestions.

I have incorporated all of them.

robertknight reviewed (approved):
LGTM.
